handlers/article.go

```diff
@@ -0,0 +1,437 @@
+package handlers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/JackyChiu/realworld-starter-kit/models"
+)
+
+type Article struct {
+	Slug           string    `json:"slug"`
+	Title          string    `json:"title"`
+	Description    string    `json:"description"`
+	Body           string    `json:"body"`
+	Favorited      bool      `json:"favorited"`
+	FavoritesCount int       `json:"favoritesCount"`
+	TagsList       []string  `json:"tagsList"`
+	CreatedAt      time.Time `json:"createdAt"`
+	UpdatedAt      time.Time `json:"updatedAt"`
+	Author         struct {
+		Username  string `json:"username"`
+		Bio       string `json:"bio"`
+		Image     string `json:"image"`
+		Following bool   `json:"following"`
+	} `json:"user"`
+}
+
+type ArticleJSON struct {
+	Article `json:"article"`
+}
+
+type ArticlesJSON struct {
+	Articles      []Article `json:"articles"`
+	ArticlesCount int       `json:"articlesCount"`
+}
+
+type errorResponse struct {
+	Errors map[string]interface{} `json:"errors"`
+}
+type Author struct {
+	Username  string `json:"username"`
+	Bio       string `json:"bio"`
+	Image     string `json:"image"`
+	Following bool   `json:"following"`
+}
+
+const (
+	contextKeyCurrentUser = contextKey("current_user")
+	contextKeyArticle     = contextKey("article")
+	contextKeyLoggedIn    = contextKey("logged_in")
+)
+
+// ArticlesHandler handle /api/articles
+func (h *Handler) ArticlesHandler(w http.ResponseWriter, r *http.Request) {
+	//h.Logger.Println(r.Method, r.URL.Path)
+
+	// Unprotected routes
+	router := NewRouter(h.Logger)
+	router.AddRoute(
+		`articles\/?$`,
+		"GET", h.getCurrentUser(http.HandlerFunc(h.getArticles)))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"GET", h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.getArticle))))
+
+	// Protected routes
+	router.AddRoute(
+		`articles\/?$`,
+		"POST", h.authorize(h.getCurrentUser(http.HandlerFunc(h.createArticle))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"PUT", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.updateArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)$`,
+		"DELETE", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.deleteArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)\/favorite$`,
+		"POST", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.favoriteArticle)))))
+
+	router.AddRoute(
+		`articles\/(?P<slug>[0-9a-zA-Z\-]+)\/favorite$`,
+		"DELETE", h.authorize(h.getCurrentUser(h.extractArticle(http.HandlerFunc(h.unFavoriteArticle)))))
+
+	//router.DebugMode(true)
+
+	router.ServeHTTP(w, r)
+}
+
+func (h *Handler) getCurrentUser(next http.Handler) http.Handler {
```

Comment:
I think the handlers and middleware should be taking and returning `http.HandlerFunc`s instead, since we are only utilizing the `ServeHTTP` in the router